src/Catalyst.Core.Modules.Dfs/BlockExchange/Bitswap.cs

```diff
@@ -0,0 +1,489 @@
+        /// </summary>
+        public Bitswap()
+        {
+            Protocols = new IBitswapProtocol[]
+            {
+                new Bitswap11 {Bitswap = this},
+                new Bitswap1 {Bitswap = this}
+            };
+        }
+
+        /// <summary>
+        ///   Provides access to other peers.
+        /// </summary>
+        public Swarm Swarm { get; set; }
+
+        /// <summary>
+        ///   Provides access to blocks of data.
+        /// </summary>
+        public IBlockApi BlockService { get; set; }
+
+        /// <summary>
+        ///   Statistics on the bitswap component.
+        /// </summary>
+        /// <seealso cref="IStatsApi"/>
+        public BitswapData Statistics
+        {
+            get
+            {
+                return new BitswapData
+                {
+                    BlocksReceived = BlocksReceived,
+                    BlocksSent = BlocksSent,
+                    DataReceived = DataReceived,
+                    DataSent = DataSent,
+                    DupBlksReceived = DupBlksReceived,
+                    DupDataReceived = DupDataReceived,
+                    ProvideBufLen = 0, // TODO: Unknown meaning
+                    Peers = Swarm.KnownPeers.Select(p => p.Id),
+                    Wantlist = wants.Keys
+                };
+            }
+        }
+
+        /// <summary>
+        ///   Gets the bitswap ledger for the specified peer.
+        /// </summary>
+        /// <param name="peer">
+        ///   The peer to get information on.  If the peer is unknown, then a ledger
+        ///   with zeros is returned.
+        /// </param>
+        /// <returns>
+        ///   Statistics on the bitswap blocks exchanged with the peer.
+        /// </returns>
+        /// <seealso cref="IBitswapApi.LedgerAsync(Lib.P2P.Peer,System.Threading.CancellationToken)"/>
+        public BitswapLedger PeerLedger(Peer peer)
+        {
+            if (peerLedgers.TryGetValue(peer, out var ledger)) return ledger;
+            return new BitswapLedger {Peer = peer};
+        }
+
+        /// <summary>
+        ///   Raised when a blocked is needed.
+        /// </summary>
+        /// <remarks>
+        ///   Only raised when a block is first requested.
+        /// </remarks>
+        public event EventHandler<CidEventArgs> BlockNeeded;
+
+        /// <inheritdoc />
+        public Task StartAsync()
+        {
+            log.Debug("Starting");
+
+            foreach (var protocol in Protocols) Swarm.AddProtocol(protocol);
+            Swarm.ConnectionEstablished += Swarm_ConnectionEstablished;
+
+            // TODO: clear the stats.
+            peerLedgers.Clear();
+
+            return Task.CompletedTask;
+        }
+
+        // When a connection is established
+        // (1) Send the local peer's want list to the remote
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        private async void Swarm_ConnectionEstablished(object sender, PeerConnection connection)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this unused method parameter 'sender'.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)